src/Hosting/Hosting/src/Internal/WebHostLifetime.cs

```diff
@@ -9,22 +11,22 @@ internal sealed class WebHostLifetime : IDisposable
     private readonly ManualResetEventSlim _resetEvent;
     private readonly string _shutdownMessage;
 
+    private PosixSignalRegistration _sigIntRegistration;
+    private PosixSignalRegistration _sigQuitRegistration;
+    private PosixSignalRegistration _sigTermRegistration;
+
     private bool _disposed;
-    private bool _exitedGracefully;
 
     public WebHostLifetime(CancellationTokenSource cts, ManualResetEventSlim resetEvent, string shutdownMessage)
     {
         _cts = cts;
         _resetEvent = resetEvent;
         _shutdownMessage = shutdownMessage;
 
-        AppDomain.CurrentDomain.ProcessExit += ProcessExit;
-        Console.CancelKeyPress += CancelKeyPress;
-    }
-
-    internal void SetExitedGracefully()
-    {
-        _exitedGracefully = true;
+        Action<PosixSignalContext> handler = HandlePosixSignal;
+        _sigIntRegistration = PosixSignalRegistration.Create(PosixSignal.SIGINT, handler);
+        _sigQuitRegistration = PosixSignalRegistration.Create(PosixSignal.SIGQUIT, handler);
+        _sigTermRegistration = PosixSignalRegistration.Create(PosixSignal.SIGTERM, handler);
```

Comment:
I think we need to handle Windows too here

Reply:
On Windows, `PosixSignalRegistration` wraps Windows equivalent of unix signals.